src/Compilers/VisualBasic/Portable/Symbols/Tuples/TupleTypeSymbol.vb

```diff
@@ -0,0 +1,987 @@
+        End Function
+
+        Private Shared Function NumberOfValueTuples(numElements As Integer, <Out()> ByRef remainder As Integer) As Integer
+            remainder = (numElements - 1) Mod (RestPosition - 1) + 1
+            Return (numElements - 1) \ (RestPosition - 1) + 1
+        End Function
+
+        Private Shared Function GetTupleUnderlyingType(elementTypes As ImmutableArray(Of TypeSymbol), syntax As VisualBasicSyntaxNode, compilation As VisualBasicCompilation, diagnostics As DiagnosticBag) As NamedTypeSymbol
+            Dim numElements As Integer = elementTypes.Length
+            Dim remainder As Integer
+            Dim chainLength As Integer = TupleTypeSymbol.NumberOfValueTuples(numElements, remainder)
+
+            Dim wellKnownType As NamedTypeSymbol = compilation.GetWellKnownType(TupleTypeSymbol.GetTupleType(remainder))
+
+            If diagnostics IsNot Nothing AndAlso syntax IsNot Nothing Then
+                Binder.ReportUseSiteError(diagnostics, syntax, wellKnownType)
+            End If
+
+            Dim namedTypeSymbol As NamedTypeSymbol = wellKnownType.Construct(ImmutableArray.Create(Of TypeSymbol)(elementTypes, (chainLength - 1) * (TupleTypeSymbol.RestPosition - 1), remainder))
+            Dim [loop] As Integer = chainLength - 1
+            If [loop] > 0 Then
+                Dim wellKnownType2 As NamedTypeSymbol = compilation.GetWellKnownType(TupleTypeSymbol.GetTupleType(TupleTypeSymbol.RestPosition))
+
+                If diagnostics IsNot Nothing AndAlso syntax IsNot Nothing Then
+                    Binder.ReportUseSiteError(diagnostics, syntax, wellKnownType2)
+                End If
+                Do
+                    Dim typeArguments As ImmutableArray(Of TypeSymbol) = ImmutableArray.Create(Of TypeSymbol)(elementTypes, ([loop] - 1) * (TupleTypeSymbol.RestPosition - 1), TupleTypeSymbol.RestPosition - 1).Add(namedTypeSymbol)
+                    namedTypeSymbol = wellKnownType2.Construct(typeArguments)
+                    [loop] -= 1
+                Loop While [loop] > 0
+            End If
+            Return namedTypeSymbol
+        End Function
+
+        Friend Shared Sub VerifyTupleTypePresent(cardinality As Integer, syntax As VisualBasicSyntaxNode, compilation As VisualBasicCompilation, diagnostics As DiagnosticBag)
+            Debug.Assert(diagnostics IsNot Nothing AndAlso syntax IsNot Nothing)
+            Dim arity As Integer
+            Dim num As Integer = TupleTypeSymbol.NumberOfValueTuples(cardinality, arity)
+            Dim wellKnownType As NamedTypeSymbol = compilation.GetWellKnownType(TupleTypeSymbol.GetTupleType(arity))
+            Binder.ReportUseSiteError(diagnostics, syntax, wellKnownType)
+
+            If num > 1 Then
+                Dim wellKnownType2 As NamedTypeSymbol = compilation.GetWellKnownType(TupleTypeSymbol.GetTupleType(TupleTypeSymbol.RestPosition))
+                Binder.ReportUseSiteError(diagnostics, syntax, wellKnownType2)
```

Comment:
Consider extracting a `GetWellKnownTupleType(position As Integer, syntaxOpt As SyntaxNode, diagnosticsOpt As DiagnosticBag)` helper method, for use here and in `GetTupleUnderlyingType` above.